server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4173,7 +4173,18 @@ public String toString() {
         return "IndexShard(shardRouting=" + shardRouting + ")";
     }
 
+    /**
+     * @deprecated use {@link #waitForPrimaryTermAndGeneration(long, long, ActionListener)} instead.
+     */
+    @Deprecated
     public void waitForSegmentGeneration(long segmentGeneration, ActionListener<Long> listener) {
-        getEngine().addSegmentGenerationListener(segmentGeneration, listener);
+        waitForPrimaryTermAndGeneration(getOperationPrimaryTerm(), segmentGeneration, listener);
```

Comment:
We have to use `operationPrimaryTerm` here instead of unknown primary term in order to use it to compare with the current primary term in `SearchEngine`. #99752 will pass the primary term from the index shard.

Reply:
I'm not sure I follow this comment? But in general, I am not sure why we deprecate and keep the old versions of waitForSegmentGeneration() and addSegmentGenerationListener(), since https://github.com/elastic/elasticsearch/pull/99752/ will not use them? We can simply remove them and merge this along with https://github.com/elastic/elasticsearch/pull/99752/ . Else, please remember to remove them after merging https://github.com/elastic/elasticsearch/pull/99752/ .
(except if I am missing something)

Reply:
The idea was to make reviewing this PR simpler by deprecating the old methods in order to make a clean stateless PR and then removing the old methods in #101471 99752